### PR TITLE
AMDGPU: Remove flat/global fmin/fmax intrinsics

### PR DESCRIPTION
**Pull request by arsenm**
These have been replaced with atomicrmw

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/105642?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#105642** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/105642?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#97051** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/97051?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#97050** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/97050?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96876** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96876?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96875** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96875?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96874** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96874?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96873** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96873?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96872** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96872?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96760** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96760?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96759** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96759?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96444** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96443** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96443?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#96442** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/96442?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95930** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95930?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95929** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95929?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by rovka — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/GlobalISel/fp64-atomics-gfx90a.ll [hunk 1]
```diff
@@ -1015,52 +1011,6 @@ main_body:
   ret void
 }
 
-define amdgpu_kernel void @global_atomic_fmin_f64_noret(ptr addrspace(1) %ptr, double %data) {
```
[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/fp64-atomics-gfx90a.ll:1018]: Yes

**Comment by doru1004** [comment]
@arsenm can you provide an alternative to the current functionality which you are removing?

**Comment by arsenm** [comment]
> @arsenm can you provide an alternative to the current functionality which you are removing? 

Use atomicrmw with appropriate metadata. See the autoupgrade code

**Review by doru1004 — CHANGES_REQUESTED** [review]
Can you provide a similar intrinsic which can be called as Intrinsic::int_amdgcn_global_atomic_fmax which calls your desired intrinsic instead?

**Comment by arsenm** [comment]
> Can you provide a similar intrinsic which can be called as Intrinsic::int_amdgcn_global_atomic_fmax which calls your desired intrinsic instead?

No. The point is no intrinsics. Adding an identical intrinsic helps nothing. Only atomicrmw is an acceptable atomic representation

**Comment by doru1004** [comment]
> > @arsenm can you provide an alternative to the current functionality which you are removing?
> 
> Use atomicrmw with appropriate metadata. See the autoupgrade code

Are you exposing something like Intrinsic::atomicrmw which I can call from Clang code gen?

**Comment by arsenm** [comment]
> 
> Are you exposing something like Intrinsic::atomicrmw which I can call from Clang code gen?

atomicrmw is already a first class IR instruction. The intrinsics were always a hack for missing operations and annotations on atomicrmw

**Comment by doru1004** [comment]
> > Can you provide a similar intrinsic which can be called as Intrinsic::int_amdgcn_global_atomic_fmax which calls your desired intrinsic instead?
> 
> No. The point is no intrinsics. Adding an identical intrinsic helps nothing. Only atomicrmw is an acceptable atomic representation

I understand and you are free to only use atomicrmw under the hood. I just need to have a way to do it from Clang code gen.

**Comment by doru1004** [comment]
> > Are you exposing something like Intrinsic::atomicrmw which I can call from Clang code gen?
> 
> atomicrmw is already a first class IR instruction. The intrinsics were always a hack for missing operations and annotations on atomicrmw

Are there examples of how to emit such an instruction in clang code gen?

**Comment by arsenm** [comment]
> > > Are you exposing something like Intrinsic::atomicrmw which I can call from Clang code gen?
> 
> > 
> 
> > atomicrmw is already a first class IR instruction. The intrinsics were always a hack for missing operations and annotations on atomicrmw
> 
> 
> 
> Are there examples of how to emit such an instruction in clang code gen?

Yes. The corresponding builtins are already emitting atomicrmw with annotations.

**Comment by doru1004** [comment]
> > > > Are you exposing something like Intrinsic::atomicrmw which I can call from Clang code gen?
> > 
> > 
> > > 
> > 
> > 
> > > atomicrmw is already a first class IR instruction. The intrinsics were always a hack for missing operations and annotations on atomicrmw
> > 
> > 
> > Are there examples of how to emit such an instruction in clang code gen?
> 
> Yes. The corresponding builtins are already emitting atomicrmw with annotations.

Can you point me to where this is happening please?

**Comment by arsenm** [comment]
> Can you point me to where this is happening please?

https://github.com/llvm/llvm-project/blob/b9c2e2e3e910f8283f52c574fd8b6a7981d6cb0d/clang/lib/CodeGen/CGBuiltin.cpp#L19552

**Comment by doru1004** [comment]
@arsenm I am not ok with the removal of these intrinsics that allow direct call to some of these atomics. I think you should keep them while you still do the changes that you want with the other atomicrmw. Your change is not backwards compatible.
It is not yet clear to me if calling create atomicrmw is equivalent to calling the old intrisincs.

**Comment by arsenm** [comment]
> @arsenm I am not ok with the removal of these intrinsics that allow direct call to some of these atomics. I think you should keep them while you still do the changes that you want with the other atomicrmw. 

The changes are already completed and we are at the final point of removal. 

> Your change is not backwards compatible. It is not yet clear to me if calling create atomicrmw is equivalent to calling the old intrisincs.

The intrinsics are strictly less expressive than atomicrmw. There is no plus to using the intrinsics, and it is trivial to migrate to an atomicrmw that will produce the ISA instruction. The old intrinsics had broken scope and ordering arguments which did not work. You only need to decide whether the use case is OK with breaking remote memory, fine grained memory, or both.

**Comment by rovka** [comment]
> @arsenm I am not ok with the removal of these intrinsics that allow direct call to some of these atomics. I think you should keep them while you still do the changes that you want with the other atomicrmw. Your change is not backwards compatible. It is not yet clear to me if calling create atomicrmw is equivalent to calling the old intrisincs.

@doru1004 Can you please provide an example use case where it's not clear how to use atomicrmw to replace the old intrinsics? Then maybe we can add something to the docs or at least commit message to clarify.

**Comment by doru1004** [comment]
> > @arsenm I am not ok with the removal of these intrinsics that allow direct call to some of these atomics. I think you should keep them while you still do the changes that you want with the other atomicrmw. Your change is not backwards compatible. It is not yet clear to me if calling create atomicrmw is equivalent to calling the old intrisincs.
> 
> @doru1004 Can you please provide an example use case where it's not clear how to use atomicrmw to replace the old intrinsics? Then maybe we can add something to the docs or at least commit message to clarify.

For all the cases where the intrinsic was used i.e. fadd/fmin/fmax emitting atomic RMW without any metadata is conservative, safe but not equivalent with the behavior of the intrinsic. So one thing that has not been clear is what is the correct combination of metadata attributes that allows the newly emitted atomic RMW instruction to be equivalent (or very very close) to the behavior of the intrinsic that was removed. I would argue that some higher level functions like emitAtomicRMW{FAdd/FMin/FMax} are needed to cover those cases and emit the atomicRMW with the correct combination of metadata attributes.

**Review by shiltian — APPROVED** [review]
LGTM, as long as @doru1004 's concern has been resolved.

**Comment by doru1004** [comment]
I would still like to see something like the functions I suggested in my previous comment. Something that will be equivalent with the intrinsic that is being eliminated. These functions would use the new atomicRMW method AND also set the attributes accordingly such that it matches the behavior of the intrinsic.

**Comment by arsenm** [comment]
> I would still like to see something like the functions I suggested in my previous comment. Something that will be equivalent with the intrinsic that is being eliminated. These functions would use the new atomicRMW method AND also set the attributes accordingly such that it matches the behavior of the intrinsic.

There won't be a new function, but @yxsamliu is working on source level controls for these fields for the atomic builtins

**Comment by doru1004** [comment]
> There won't be a new function, but @yxsamliu is working on source level controls for these fields for the atomic builtins

I need to be able to call this from Clang code gen. The reason I insist on this is because I don't think the selection of attributes to pass to the current atomicRMW function is clear enough i.e. if I was relying on the intrinsic up until now, it is not clear what set of additional options to pass to atomicRMW() to make it behave like the intrinsic. So I would like that knowledge to live somewhere in the compiler i.e. in the shape of a function atomicRMWFastFAdd or something like that which calls atomicRMW under the hood BUT ALSO provides the set of additional options that make it behave like the intrinsic.

**Comment by arsenm** [comment]
> I need to be able to call this from Clang code gen. The reason I insist on this is because I don't think the selection of attributes to pass to the current atomicRMW function is clear enough i.e. if I was relying on the intrinsic up until now, it is not clear what set of additional options to pass to atomicRMW() to make it behave like the intrinsic. 

This is exactly why we must remove this intrinsic or anything that acts like it. You, the frontend writer, must semantically know what you can do. The semantics cannot be "do whatever the intrinsic happens to do on this target"

**Comment by doru1004** [comment]
> This is exactly why we must remove this intrinsic or anything that acts like it. You, the frontend writer, must semantically know what you can do. The semantics cannot be "do whatever the intrinsic happens to do on this target"

But you're not removing everything that acts like it except for the intrinsic itself. There's nothing illegal about asking for a new style of atomic RMW with a particular set of additional options that happens to match the previous intrinsic.

**Comment by arsenm** [comment]
> But you're not removing everything that acts like it except for the intrinsic itself. There's nothing illegal about asking for a new style of atomic RMW with a particular set of additional options that happens to match the previous intrinsic.

Part of the point of this is to make you do this work. Adding a convenience to "match the intrinsic" is counterproductive. The intrinsic was not semantically defined

**Comment by doru1004** [comment]
> Part of the point of this is to make you do this work. Adding a convenience to "match the intrinsic" is counterproductive. The intrinsic was not semantically defined

The problem is that no one knows what options to use to make the atomicRMW closer to the old intrinsic. You do not document that anywhere.

**Comment by arsenm** [comment]
> The problem is that no one knows what options to use to make the atomicRMW closer to the old intrinsic. You do not document that anywhere.

[It's in AMDGPUUsage](https://llvm.org/docs/AMDGPUUsage.html#llvm-ir-metadata)

**Comment by doru1004** [comment]
> > The problem is that no one knows what options to use to make the atomicRMW closer to the old intrinsic. You do not document that anywhere.
> 
> [It's in AMDGPUUsage](https://llvm.org/docs/AMDGPUUsage.html#llvm-ir-metadata)

I agree but that doesn't say what combination of those options to use to be as close as possible to the previous intrinsic. So if we can document that then we will be all good!

**Pull request merged**